src/cir.rs

```diff
@@ -0,0 +1,439 @@
+                    } else {
+                        let next_next_ins = &lir.instructions[index + 2];
+                        let or_ins = OrIns {
+                            left: CirOperand::Predicate(next_ins.as_predicate().unwrap().clone()),
+                            right: CirOperand::Predicate(
+                                next_next_ins.as_predicate().unwrap().clone(),
+                            ),
+                        };
+                        cir.instructions.push(CirInstruction::OrIns(or_ins));
+                        operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
+                        index += 3;
+
+                        reduce_translation_stack(
+                            &mut cir.instructions,
+                            &mut operator_stack,
+                            &mut operand_stack,
+                        );
+                    }
+                }
+                LirLogicalOperators::Not => {
+                    let next_ins = &lir.instructions[index + 1];
+                    if is_operator(next_ins) {
+                        // push LIR operator to ops stack
+                        // next is operator
+                        operator_stack.push(LirLogicalOperators::Not);
+                        index += 1;
+                    } else {
+                        next_ins.as_predicate().unwrap(); //right
+                        let not_ins = NotIns {
+                            right: CirOperand::Predicate(next_ins.as_predicate().unwrap().clone()),
+                        };
+                        cir.instructions.push(CirInstruction::NotIns(not_ins));
+                        operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
+                        index += 2;
+
+                        reduce_translation_stack(
+                            &mut cir.instructions,
+                            &mut operator_stack,
+                            &mut operand_stack,
+                        );
+                    }
+                }
+            },
+            LirInstruction::Predicate(p) => {
+                operand_stack.push(CirOperand::Predicate(p.clone()));
+                reduce_translation_stack(
+                    &mut cir.instructions,
+                    &mut operator_stack,
+                    &mut operand_stack,
+                );
+
+                index += 1;
+            }
+        }
+
+        if index >= lir.instructions.len() {
+            // end of LirProgram
+            break;
+        }
+    }
+    debug_assert_eq!(operator_stack.len(), 0);
+}
+
+fn execute_helper(
+    cir_instructions: &[CirInstruction],
+    index: usize,
+    ctx: &mut Context,
+    m: &mut Match,
+) -> bool {
+    match &cir_instructions[index] {
+        CirInstruction::AndIns(and) => {
+            let left_val = if is_index(&and.left) {
+                execute_helper(cir_instructions, and.left.as_index().unwrap(), ctx, m)
```

Comment:
Does The deepest recursion depth here as same as the previous `Expression::execution` if no short circuit happens here?

Reply:
the depth should be same.